### PR TITLE
feat: add validation loop

### PR DESCRIPTION
**Pull request by hbredin**
This will allow the use of learning rate schedulers such as ReduceLROnPlateau.
https://pytorch-lightning.readthedocs.io/en/latest/optimizers.html#learning-rate-scheduling

Also factorizes segmentation tasks codebase as there was lots of redundant code between them.

### CODE REVIEW

**Review by mogwai — COMMENTED** [review]
This is great, you've removed a lot of code here :+1:

**Review by mogwai — APPROVED** [review]
